### ISSUE TITLE
[Question] Webdataset numpy decoding sizes

### ISSUE THREAD

**Issue by austinmw** [NVIDIA/DALI#3690]
Hi, I created a webdataset where I've encoded images as jpgs and labels as padded numpy arrays of size `(64, 5)` (where the 5 represents `[x,y,w,h,id]`).

When I attempt to decode using `fn.readers.webdataset` and `ext=['jpg', 'npy']`,  I'm getting label sizes that I don't understand. They all seem to end in `8`. For example, this pipeline produces the below output for the first several steps:

```python
def _decode_augment(img, seed=0):
    img = fn.decoders.image_random_crop(img, device='mixed', output_type=types.RGB, 
                                        random_aspect_ratio=1.0, random_area=1.0)
    img = fn.resize(img, device='gpu', size=(416,416), dtype=types.FLOAT)
    return img

@pipeline_def(batch_size=1, num_threads=num_workers, device_id=0)
def wds_pipeline(urls, seed=0):
    images, labels = fn.readers.webdataset(
        paths=urls,
        ext=['jpg', 'npy'],
        missing_component_behavior='error',
        random_shuffle=True,
        seed=seed,
        name="Reader",
    )
    return _decode_augment(images, seed=seed), labels

train_pipe = wds_pipeline(urls=dali_urls)

train_loader = DALIGenericIterator(
    train_pipe,
    ["images", "labels"],
    reader_name="Reader",
    last_batch_policy=LastBatchPolicy.DROP)

for item in train_loader:
    images = item[0]['images']
    labels = item[0]['labels']
    print(labels.shape)
```

```
torch.Size([1, 208])
torch.Size([1, 328])
torch.Size([1, 368])
torch.Size([1, 168])
torch.Size([1, 208])
torch.Size([1, 768])
torch.Size([1, 208])
torch.Size([1, 208])
torch.Size([1, 448])
torch.Size([1, 448])
torch.Size([1, 248])
...
```

**Comment by JanuszL** [comment]
Hi @austinmw,

As far as I know, Webdataset when asked to encode data as a NumPy array creates entries in NumPy file format. That means NumPy header and content. DALI threats this data as an u8 array. I think you should save this data as a raw sequence of bytes:
```
label= x.to_bytes(32, 'little') + y.to_bytes(32, 'little') + w.to_bytes(32, 'little') + h.to_bytes(32, 'little') + id.to_bytes(32, 'little')
label = label.to_bytes(8, 'little')
            sample = {'__key__': key, 'jpg': image, 'cls': label}

...

images, labels = fn.readers.webdataset(
        paths=urls,
        ext=['jpg', 'npy'],
        missing_component_behavior='error',
        random_shuffle=True,
        seed=seed,
        name="Reader",
        dtypes=[types.UINT8, types.INT32]
    )
```

**Comment by austinmw** [comment]
Thanks!

**Comment by austinmw** [comment]
Hey @JanuszL, when I read in like this, my `bboxes` and `labels` end up 1-dimensional. How can I then resize them before passing them to `fn.random_bbox_crop(bboxes, labels, ...)`? (Or alternatively, is there a way to read them in directly with the correct dimensions?)

**Comment by JanuszL** [comment]
I think you should use [reshape operator](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#nvidia.dali.fn.reshape), not resize (unless it is a typo).

**Comment by evgeniishch** [comment]
Hello there @JanuszL ! I came across this issue recently, trying to use `dali.fn.readers.webdataset` reader. 
It works OK with images (which are basically integer arrays), however I struggle with encoding labels which in my case are **float arrays**, because, as mentioned above, 
> DALI treats this data as an u8 array

 Is there any elegant/painless way to write numpy/torch float array to webdataset and then read it with `dali.fn.readers.webdataset`? The pipeline i am building looks pretty much like this: 
```python
# -------------------- CREATE WEBDATASET --------------------
import webdataset as wds
import numpy as np
import torch

sink = wds.TarWriter("WDS.tar")

sample = {}
sample["__key__"] = "example"

img_path = 'path/to/img.jpg'
with open(img_path, 'rb') as stream:
    img = stream.read()
sample[f"img.jpg"] = img

label = torch.rand(8, 2, dtype=torch.float32)
sample['labels.npy'] = label.numpy().tobytes() # How shall I encode labels and what dict key shall I use ?

# -------------------- DALLI PART --------------------
import nvidia.dali as dali
from nvidia.dali import pipeline_def
import nvidia.dali.fn as fn
import nvidia.dali.types as types
from nvidia.dali.plugin.pytorch import DALIGenericIterator, LastBatchPolicy

@pipeline_def
def create_dali_pipeline(wds_path, device, shard_id=0, num_shards=1):
    img, labels = fn.readers.webdataset(
        paths=wds_path, 
        ext=[
            "img.jpg", 
            "labels.npy", 
        ], 
        dtypes=[
            types.DALIDataType.UINT8, 
            types.DALIDataType.UINT8, # which type shall i use here?
        ],
        missing_component_behavior="error"
    )

    # some preprocssing here

    return img, labels

pipeline = create_dali_pipeline(
            wds_path='WDS.tar',
            batch_size=16,
            device="gpu",
            device_id=0,
            shard_id=0,
            num_shards=2,
            num_threads=8,
        )

train_loader = DALIGenericIterator(
            pipeline, 
            ['imgs', 'labels'], 
            last_batch_policy=LastBatchPolicy.PARTIAL
)

sample = next(iter(train_loader))[0]
print(f"Labels: {sample['labels']}")
```

**Comment by JanuszL** [comment]
Hi @evgeniishch.

It should be enough to:
```
import webdataset as wds
import numpy as np
import torch

sink = wds.TarWriter("WDS.tar")

sample = {}
sample["__key__"] = "example"

img_path = 'path/to/img.jpg'
with open(img_path, 'rb') as stream:
    img = stream.read()
sample[f"img.jpg"] = img

label = torch.rand(8, 2, dtype=torch.float32)
sample['labels.npy'] = label.numpy().tobytes()
print(label)
sink.write(sample)
sink.close()

# -------------------- DALI PART --------------------
import nvidia.dali as dali
from nvidia.dali import pipeline_def
import nvidia.dali.fn as fn
import nvidia.dali.types as types
from nvidia.dali.plugin.pytorch import DALIGenericIterator, LastBatchPolicy

@pipeline_def
def create_dali_pipeline(wds_path, device, shard_id=0, num_shards=1):
    img, labels = fn.readers.webdataset(
        paths=wds_path,
        ext=[
            "img.jpg",
            "labels.npy",
        ],
        dtypes=[
            types.DALIDataType.UINT8,
            types.DALIDataType.FLOAT,
        ],
        missing_component_behavior="error"
    )

    # some preprocssing here

    return img, labels

pipeline = create_dali_pipeline(
            wds_path='WDS.tar',
            batch_size=16,
            device="gpu",
            device_id=0,
            shard_id=0,
            num_shards=2,
            num_threads=8,
        )

train_loader = DALIGenericIterator(
            pipeline,
            ['imgs', 'labels'],
            last_batch_policy=LastBatchPolicy.PARTIAL
)

sample = next(iter(train_loader))[0]
print(f"imgs: {sample['imgs']}")
print(f"Labels: {sample['labels']}")```
However you need to store shape as well. Now they are stored just as an array of data.

**Comment by evgeniishch** [comment]
@JanuszL Thank you! That was really helpful!
